File: web/blue/article.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
@Author: reber
@Mail: reber0ask@qq.com
@Date: 2020-12-17 17:03:50
LastEditTime: 2021-09-16 11:21:51
'''

import re
import demjson
import requests
from urllib.parse import urlparse, urlunparse
from flask import Blueprint
from flask import request
from flask import jsonify
from flask import make_response
from sqlalchemy import func

from sqlmodule import session_maker
from sqlmodule import Config
from sqlmodule import Article
from sqlmodule import Data
from sqlmodule import to_dict
from sqlmodule import utc2bj

from lib.data import global_data
from lib.request import req

from .common import is_login
from .common import get_user_id
from .common import get_user_role
from .common import logger_user_action


article_blueprint = Blueprint('article', __name__)


@article_blueprint.route("/check", methods=['POST'])
@is_login
@logger_user_action(msg_type="user")
def check_regex():
    data = request.get_json()
    name = data.get("name", "")
    link = data.get("link", "")
    regex = data.get("regex", "")

    o = urlparse(link)
    base_url = urlunparse((o.scheme, o.netloc, "/", "", "", ""))

    article_tag = list()
    try:
        resp = req.get(url=link)
        href_text_list = re.findall(regex, str(resp.content, encoding='utf-8'), re.S|re.M)
        if href_text_list:
            for href_text in href_text_list:
                a_href = href_text[0].strip()
                if not a_href.startswith("http"):
                    a_href = base_url+a_href.lstrip("/")
                a_text = href_text[1].strip()
                article_tag.append({"title": a_text, "url": a_href})
    except requests.exceptions.ConnectionError as e:
        r_data = {"code": 1, "msg": "requests.exceptions.ConnectionError"}
        return make_response(jsonify(r_data), 500)
    except Exception as e:
        global_data.logger.error(str(e))
        r_data = {"code": 1, "msg": "server error"}
        return make_response(jsonify(r_data), 500)
    else:
        r_data = {"code": 0, "msg": "", "data": article_tag[:5], "count": len(article_tag)}
        return make_response(jsonify(r_data), 200)


@article_blueprint.route("/add", methods=['POST'])
@is_login
@logger_user_action(msg_type="user")
def add_blog():
    post_data = request.get_json()
    name = post_data.get("name", "")
    link = post_data.get("link", "")
    regex = post_data.get("regex", "")

    access_token = request.headers.get('access_token')
    user_id = get_user_id(access_token)

    r_data = dict()
    affect_num = 0
    try:
        with session_maker(global_data.sqlite_uri) as db_session:
            result = db_session.query(Config.value).filter_by(key="website") .first()
            result = demjson.decode(result[0])
            rss_site = result.get("domain")

            site = Article(user_id=user_id, name=name, link=link, regex=regex)
            db_session.add(site)
            db_session.flush()
            db_session.refresh(site)
            ref_id = site.id

            rss = "{}/api/data/rss/{}/article/{}".format(rss_site.rstrip("/"), user_id, ref_id)
            affect_num = db_session.query(Article).filter_by(id=ref_id).update({"rss": rss})

    except Exception as e:
        global_data.logger.error(str(e))
        r_data = {"code": 1, "msg": "add site error"}
        return make_response(jsonify(r_data), 500)
    else:
        if affect_num:
            r_data = {"code": 0, "msg": "得到 rss 链接: <br>{}".format(rss)}
            return make_response(jsonify(r_data), 200)
        else:
            r_data = {"code": 1, "msg": "add site error"}
            return make_response(jsonify(r_data), 500)

@article_blueprint.route("/list", methods=['POST'])
@is_login
def list_blog():
    data = request.get_json()
    pageIndex = data.get("page", 0)
    pageSize = data.get("limit", 0)

    access_token = request.headers.get('access_token')
    user_id = get_user_id(access_token)
    user_role = get_user_role(access_token)

    r_data = dict()
    data = list()
    try:
        with session_maker(global_data.sqlite_uri) as db_session:
            count = db_session.query(func.count(Article.id)).filter(
                (Article.user_id==user_id) | (user_role=="root")).scalar()
            results = db_session.query(Article).filter(
                (Article.user_id==user_id) | (user_role=="root")).order_by(Article.id.desc()).limit(pageSize).offset((pageIndex-1)*pageSize).all()
            _data = to_dict(results)
        if _data:
            for x in _data:
                x["add_time"] = utc2bj(x["add_time"])
                data.append(x)
    except Exception as e:
        global_data.logger.error(str(e))
        r_data = {"code": 1}
        return make_response(jsonify(r_data), 500)
    else:
        r_data = {"code": 0, "msg": "", "data": data, "count": count}
        return make_response(jsonify(r_data), 200)

@article_blueprint.route("/update", methods=['POST'])
@is_login
@logger_user_action(msg_type="user")
def update_blog():
    data = request.get_json()
    _id = data.get("id", "")
    data.pop("id")
    data.pop("add_time")

    access_token = request.headers.get('access_token')
    user_id = get_user_id(access_token)
    user_role = get_user_role(access_token)

    try:
        from sqlmodule import User
        from sqlalchemy import or_, and_
        with session_maker(global_data.sqlite_uri) as db_session:
            affect_num = db_session.query(Article).filter(
                (Article.user_id==user_id) | (user_role=="root")).filter_by(id=_id).update(data, synchronize_session=False)
    except Exception as e:
        # r_data["msg"] = str(e).split('\n')[0]
        global_data.logger.error(str(e))
        r_data = {"code": 1}
        return make_response(jsonify(r_data), 500)
    else:
        if affect_num:
            r_data = {"code": 0}
            return make_response(jsonify(r_data), 200)
        else:
            r_data = {"code": 1, "msg": "Permission denied"}
            return make_response(jsonify(r_data), 403)

@article_blueprint.route("/delete", methods=['POST'])
@is_login
@logger_user_action(msg_type="user")
def delete_blog():
    data = request.get_json()
    _id = data.get("id")

    access_token = request.headers.get('access_token')
    user_id = get_user_id(access_token)
    user_role = get_user_role(access_token)

    affect_num = 0
    r_data = dict()
    try:
        with session_maker(global_data.sqlite_uri) as db_session:
            affect_num = db_session.query(Article).filter(
                (Article.user_id==user_id) | (user_role=="root")).filter(Article.id==_id).delete(synchronize_session=False)
            if affect_num:
                count = db_session.query(func.count(Data.id)).filter(Data.category=="article").filter(Data.ref_id==_id).scalar()
                if count:
                    affect_num = db_session.query(Data).filter(Data.category=="article").filter(Data.ref_id==_id).delete()
    except Exception as e:
        global_data.logger.error(str(e))
        r_data = {"code": 1, "msg": "delete error"}
        return make_response(jsonify(r_data), 500)
    else:
        if affect_num:
            r_data = {"code": 0, "msg": "delete success"}
            return make_response(jsonify(r_data), 200)
        else:
            r_data = {"code": 1, "msg": "delete error"}
            return make_response(jsonify(r_data), 403)

@article_blueprint.route("/search", methods=['POST'])
@is_login
@logger_user_action(msg_type="user")
def search_article_site():
    data = request.get_json()
    pageIndex = data.get("page", 0)
    pageSize = data.get("limit", 0)
    name = data.get("name", "")

    access_token = request.headers.get('access_token')
    user_id = get_user_id(access_token)
    user_role = get_user_role(access_token)

    r_data = dict()
    data = list()
    try:
        with session_maker(global_data.sqlite_uri) as db_session:
            count = db_session.query(func.count(Article.id)).filter(
                (Article.user_id==user_id) | (user_role=="root"), Article.name.like("%{}%".format(name))).scalar()
            results = db_session.query(Article).filter(
                (Article.user_id==user_id) | (user_role=="root"), Article.name.like("%{}%".format(name))).order_by(Article.id.desc()).limit(pageSize).offset((pageIndex-1)*pageSize).all()
            _data = to_dict(results)

        if _data:
            for x in _data:
                x["add_time"] = utc2bj(x["add_time"])
                data.append(x)
    except Exception as e:
        logger.error(str(e))
        r_data = {"code": 1}
        return make_response(jsonify(r_data), 500)
    else:
        r_data = {"code": 0, "msg": "", "data": data, "count": count}
        return make_response(jsonify(r_data), 200)

File: routers/article.go
/*
 * @Author: reber
 * @Mail: reber0ask@qq.com
 * @Date: 2022-01-04 20:52:53
 * @LastEditTime: 2022-04-30 09:50:38
 */
package routers

import (
	"RssSub/global"
	"RssSub/middleware"
	"RssSub/mydb"
	"fmt"
	"regexp"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/reber0/go-common/parse"
	"github.com/reber0/go-common/utils"
)

// Article Site 相关路由
func ArticleRouter(r *gin.Engine) {
	articleGroup := r.Group("/api/article").Use(middleware.JWTAuth(), middleware.Action())
	{
		articleGroup.POST("/check", articleCheckRegex)
		articleGroup.POST("/list", articleSiteList)
		articleGroup.POST("/add", articleSiteAdd)
		articleGroup.POST("/update", articleSiteUpdate)
		articleGroup.POST("/delete", articleSiteDelete)
		articleGroup.POST("/search", articleSiteSearch)
	}
}

func articleCheckRegex(c *gin.Context) {
	type PostData struct {
		Link  string `form:"link" json:"link"`
		Regex string `form:"regex" json:"regex"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "检查失败",
		})
	} else {
		base_url := parse.NewParseURL(postJson.Link).BaseURL()
		resp, err := global.Client.Get(postJson.Link)
		if resp != nil {
			html := resp.Html()

			article_tag := make([]map[string]string, 0, 100)
			reg := regexp.MustCompile(`(?sm)` + postJson.Regex)
			result := reg.FindAllStringSubmatch(html, -1)
			for _, href_text := range result {
				a_href := strings.TrimSpace(href_text[1])
				if !strings.HasPrefix(a_href, "http") {
					a_href = base_url + strings.TrimLeft(a_href, "/")
				}
				a_text := strings.TrimSpace(href_text[2])
				article_tag = append(article_tag, map[string]string{"title": a_text, "url": a_href})
			}

			c.JSON(200, gin.H{
				"code": 0,
				"data": article_tag[0:5],
			})
		}
		if err != nil {
			global.Log.Error(err.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "error",
			})
		}
	}
}

func articleSiteAdd(c *gin.Context) {
	type PostData struct {
		Name  string `form:"name" json:"name"`
		Link  string `form:"link" json:"link"`
		Regex string `form:"regex" json:"regex"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "检查失败",
		})
	} else {
		userId := c.GetString("uid")

		var domain string
		global.Db.Model(&mydb.Config{}).Select("value").Where("key='domain'").First(&domain)

		site := mydb.Article{UID: userId, Name: postJson.Name, Link: postJson.Link, Regex: postJson.Regex}
		result := global.Db.Create(&site)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "添加失败",
			})
			return
		}
		refId := site.ID

		rssPath := fmt.Sprintf("/api/data/rss/%s/article/%d", userId, refId)
		result = global.Db.Model(&mydb.Article{}).Where("id=?", refId).Update("rss", rssPath)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "添加失败",
			})
			return
		}

		rss := fmt.Sprintf("%s/api/data/rss/%s/article/%d", strings.TrimRight(domain, "/"), userId, refId)
		c.JSON(200, gin.H{
			"code": 0,
			"msg":  fmt.Sprintf("得到 rss 链接: <br>%s", rss),
		})
	}
}

func articleSiteList(c *gin.Context) {
	type PostData struct {
		PageIndex int `form:"page" json:"page"`
		PageSize  int `form:"limit" json:"limit"`
	}

	type RespData struct {
		ID        int    `json:"id"`
		Name      string `json:"name" gorm:"column:name; type:varchar(100); not null; comment:博客名字"`
		Link      string `json:"link" gorm:"column:link; type:varchar(100); not null; comment:文章网站的网址"`
		Regex     string `json:"regex" gorm:"column:regex; type:text; not null; comment:正则"`
		Rss       string `json:"rss" gorm:"column:rss; type:varchar(100); comment:RSS 地址"`
		CreatedAt string `json:"created_at" gorm:"column:created_at; type:varchar(100); comment:添加时间"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "查询失败",
		})
	} else {
		userId := c.GetString("uid")
		_, role := GetUserMsg(userId)

		var count int64
		var datas []RespData
		result := global.Db.Model(&mydb.Article{}).Select("id", "name", "link", "regex", "rss", "created_at").Where(
			"uid=? or ?='root'", userId, role).Order("id desc").Count(&count).Limit(postJson.PageSize).Offset((postJson.PageIndex - 1) * postJson.PageSize).Find(&datas)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "查询失败",
			})
			return
		}

		var domain string
		global.Db.Model(&mydb.Config{}).Select("value").Where("key='domain'").First(&domain)

		for index, data := range datas {
			datas[index].Rss = strings.TrimRight(domain, "/") + data.Rss
			datas[index].CreatedAt = utils.UnixToStr(data.CreatedAt)
		}

		c.JSON(200, gin.H{
			"code":  0,
			"data":  datas,
			"count": count,
		})
	}
}

func articleSiteUpdate(c *gin.Context) {
	type PostData struct {
		ID    int    `form:"id" json:"id"`
		Name  string `form:"name" json:"name"`
		Link  string `form:"link" json:"link"`
		Regex string `form:"regex" json:"regex"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "更新失败",
		})
	} else {
		userId := c.GetString("uid")
		_, role := GetUserMsg(userId)

		updateData := mydb.Article{Name: postJson.Name, Link: postJson.Link, Regex: postJson.Regex}
		result := global.Db.Model(&mydb.Article{}).Where(
			"(uid=? or ?='root') and id=?", userId, role, postJson.ID).Updates(updateData)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "更新失败",
			})
			return
		}

		c.JSON(200, gin.H{
			"code": 0,
			"msg":  "更新成功",
		})
	}
}

func articleSiteDelete(c *gin.Context) {
	type PostData struct {
		ID int `form:"id" json:"id"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "删除失败",
		})
	} else {
		userId := c.GetString("uid")
		_, role := GetUserMsg(userId)

		id := postJson.ID

		result := global.Db.Where("(uid=? or ?='root') and id=?", userId, role, id).Delete(&mydb.Article{})
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "删除失败",
			})
			return
		}
		result = global.Db.Where("category='article' and ref_id=?", id).Delete(&mydb.Data{})
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "删除失败",
			})
			return
		}

		c.JSON(200, gin.H{
			"code": 0,
			"msg":  "删除成功",
		})
	}
}

func articleSiteSearch(c *gin.Context) {
	type PostData struct {
		PageIndex int    `form:"page" json:"page"`
		PageSize  int    `form:"limit" json:"limit"`
		KeyWord   string `form:"keyword" json:"keyword"`
	}

	type RespData struct {
		ID        int    `json:"id"`
		UID       string `json:"uid" gorm:"column:uid; size:32; not null; unique; comment:用户唯一 id(uuid)"`
		Name      string `json:"name" gorm:"column:name; type:varchar(100); not null; comment:博客名字"`
		Link      string `json:"link" gorm:"column:link; type:varchar(100); not null; comment:文章网站的网址"`
		Regex     string `json:"regex" gorm:"column:regex; type:text; not null; comment:正则"`
		Rss       string `json:"rss" gorm:"column:rss; type:varchar(100); comment:RSS 地址"`
		CreatedAt string `json:"created_at" gorm:"column:created_at; type:varchar(100); comment:添加时间"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "查询失败",
		})
	} else {
		userId := c.GetString("uid")
		_, role := GetUserMsg(userId)

		keyword := fmt.Sprintf("%%%s%%", postJson.KeyWord)

		var count int64
		var datas []RespData
		result := global.Db.Model(&mydb.Article{}).Select("id", "name", "link", "regex", "rss", "created_at").Where(
			"(uid=? or ?='root') and name like ?", userId, role, keyword).Order("id desc").Count(&count).Limit(postJson.PageSize).Offset((postJson.PageIndex - 1) * postJson.PageSize).Find(&datas)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "查询失败",
			})
			return
		}

		for index, data := range datas {
			datas[index].CreatedAt = utils.UnixToStr(data.CreatedAt)
		}

		c.JSON(200, gin.H{
			"code":  0,
			"data":  datas,
			"count": count,
		})
	}
}
